shared/src/main/scala-2.10/scalajson.ast/package.scala

```diff
@@ -343,4 +345,280 @@ package object ast {
+                var i2: Int = 0
+                while (i2 < resultNegativeEFlag) {
+                  result = result / radix
+                  i2 += 1
+                }
+              } else
+                return None
+            }
+          }
+
+        val maxLenCheck = digitLength <= maxLengthConstant
+
+        if (result < multmin && maxLenCheck && !eFlag)
+          return None
+
+        if (!(digit == 0 && (decimalFlag || eFlag))) {
+
+          if (!negativeEFlag) {
+            result *= radix
+            if (result < limit + digit && maxLenCheck)
+              return None
+            result -= digit
+          }
+        }
+      }
+      i += 1
+    }
+    if (result < limit)
+      None
+    else {
+      if (negative)
+        Some(result.toLong)
+      else
+        Some(-result.toLong)
+    }
+  }
+
+  private[ast] def toBigInt(value: String): Option[BigInt] = {
+    var decimalFlag = false
+    var result: BigInt = 0
+    var negative = false
+    var char: Char = 0
+    var i = 0
+    var eFlag = false
+    var trailingZeroes: BigInt = 0
+    var negativeEFlag: Boolean = false
+    var resultNegativeEFlag: BigInt = 0
+    var digitLength: BigInt = 0
+    if (value.charAt(0) == '-') {
+      negative = true
+      i += 1
+    }
+
+    while (i < value.length) {
+      char = value.charAt(i)
+      if (char == '.')
+        decimalFlag = true
+      else if ((char | 0x20) == 'e') {
+        eFlag = true
+        val charNext = value.charAt(i + 1)
+        if (charNext == '-')
+          negativeEFlag = true
+        if (negativeEFlag || charNext == '+') {
+          i += 1
+        }
+      } else {
+        if (!(eFlag || decimalFlag))
+          digitLength += 1
+
+        val digit = Character.digit(char, radix)
+
+        if (digit == 0)
+          if (!decimalFlag)
+            trailingZeroes += 1
+          else if (!decimalFlag)
+            trailingZeroes = 0
+
+        if (digit != 0 && (decimalFlag || eFlag))
+          if (!negativeEFlag)
+            return None
+          else {
+            if (trailingZeroes != 0) {
+              resultNegativeEFlag *= radix
+              resultNegativeEFlag += digit
+              if (trailingZeroes >= resultNegativeEFlag) {
+                var i2: Int = 0
+                while (i2 < resultNegativeEFlag) {
+                  result = result / radix
+                  i2 += 1
+                }
+              } else
+                return None
+            }
+          }
+
+        if (!(digit == 0 && (decimalFlag || eFlag))) {
+
+          if (!negativeEFlag) {
+            result *= radix
+            result -= digit
+          }
+        }
+      }
+      i += 1
+    }
+    if (negative)
+      Some(result)
+    else
+      Some(-result)
+  }
+
+  private[ast] def toDouble(value: String): Option[Double] = {
+    try {
+      val asDouble = value.toDouble
+      if (BigDecimal(value, MathContext.UNLIMITED) == BigDecimal(asDouble, MathContext.UNLIMITED))
```

Comment:
Well actually, this solution (which is taken from argonaut) is consistent with how the other methods behave.

1. It will detect underflow or overflow, because the `BigDecimal` constructor will fail
2. It will also detect approximation issues, because when you construct a `BigDecimal` out of a raw string value it will not approximate the value (it uses real precision or it fails) and then we compare this to the `value.toString`)

> Number can be represented, but possibly inexactly. Return Some(value.toDouble)

Actually in this case I believe we should still return `None`, else its not consistent with the other `toInt`/`toBigInt`/`toBigDecimal` functions, which do not return inexact results. Either we do not loss any data (precision, rounding or inexact representations). Otherwise if you make this argument for `Double`, then you can argue that `JNumber("1234.01").toInt` should return `Some(1234)` (since its also an approximation).

Either we need to be consistent, or we don't. This is the same reason why `.toFloat` isn't there.

In regards to performance, if there is a more performant implementation I would love to use it. I am not sure if I have time to implement it correctly myself, considering that this is probably the last outstanding issue